src/Microsoft.AspNet.OData.Shared/Builder/ODataConventionModelBuilder.cs

```diff
@@ -699,6 +703,15 @@ private PropertyKind GetPropertyType(PropertyInfo property, out bool isCollectio
                 return PropertyKind.Dynamic;
             }
 
+            // IDictionary<string, IDictionary<string, object>> is used as a container to save/retrieve instance annotation properties for a CLR type.
+            // It is different from other collections (for example, IDictionary<string,IDictionary<string, int>>)          
+            if (typeof(IDictionary<string, IDictionary<string, object>>).IsAssignableFrom(property.PropertyType))
```

Comment:
>typeof(IDictionary<string, IDictionary<string, object>>) [](start = 16, length = 56)

so any property of type IDictionary<string,IDictionary<string,object>> is assumed to be for storing instance annotations? What if there are multiple such properties?

Reply:
As discussed keeping it

Reply:
One thing to consider is to introduce your own interface (and associated default implementation):
```
public interface IInstanceAnnotations : IDictionary<string, IDictionary<string, object>>
{
}
```

That way you get can detect your scenario explicitly and not run into ambiguity issues. We've had similar issues with the auto-mapping of `IDictionary<string, object>` for dynamic properties, and had to explicitly exclude the the dictionaries from our model.

If we did it this way would we still want a nested dictionary, or would `IDictionary<string, object>` be ideal for annotations as well (where keys would be like `property@namespace.annotation` and `namespace.annotation`)?

Taking a step further back, should we avoid using properties to detect "object level" state and instead define `IAnnotatable` and `IOpenType` that can be implemented by a model class in order to opt-in to the corresponding functionality?